curvefs/src/mds/fs_manager.cpp

```diff
@@ -0,0 +1,424 @@
+                        << ", ret = " << FSStatusCode_Name(ret);
+                return ret;
+            }
+            break;
+        case FsStatus::DELETING:
+            LOG(WARNING) << "DeleteFs already in deleting, fsName = " << fsName;
+            break;
+        default:
+            LOG(ERROR) << "DeleteFs fs in wrong status, fsName = " << fsName
+                       << ", fs status = " << FsStatus_Name(status);
+            return FSStatusCode::UNKNOWN_ERROR;
+    }
+
+    // 4. use metaserver interface, delete inode and dentry, todo
+    ret = CleanFsInodeAndDentry(mdsFsInfo.GetFsId());
+    if (ret != FSStatusCode::OK) {
+        LOG(ERROR) << "DeleteFs fail,clean inode and dentry fail"
+                        << ", fsName = " << fsName
+                        << ", ret = " << FSStatusCode_Name(ret);
+        return ret;
+    }
+
+    // 5. use space interface, destory space, todo
+
+    // 6. delete fs
+    ret = fsStorage_->Delete(fsName);
+    if (ret != FSStatusCode::OK) {
+        LOG(WARNING) << "DeleteFs fail, delete fs fail, fsName = " << fsName
+                     << ", errCode = " << ret;
+        return ret;
+    }
+
+    return FSStatusCode::OK;
+}
+
+FSStatusCode FsManager::MountFs(const std::string& fsName,
+                        const MountPoint& mountpoint, FsInfo* fsInfo) {
+    // 1. query fs
+    MdsFsInfo mdsFsInfo;
+    FSStatusCode ret = fsStorage_->Get(fsName, &mdsFsInfo);
+    if (ret != FSStatusCode::OK) {
+        LOG(WARNING) << "MountFs fail, get fs fail, fsName = " << fsName
+                  << ", errCode = " << FSStatusCode_Name(ret);
+        return ret;
+    }
+
+    // 2. check fs status
+    FsStatus status = mdsFsInfo.GetStatus();
+    switch (status) {
+        case FsStatus::NEW:
+            LOG(WARNING) << "MountFs fs is not inited, fsName = " << fsName;
+            return FSStatusCode::NOT_INITED;
+        case FsStatus::INITED:
+            // inited status, go on process
+            break;
+        case FsStatus::DELETING:
+            LOG(WARNING) << "MountFs fs is in deleting, fsName = " << fsName;
+            return FSStatusCode::UNDER_DELETRING;
+        default:
+            LOG(ERROR) << "MountFs fs in wrong status, fsName = " << fsName
+                       << ", fs status = " << FsStatus_Name(status);
+            return FSStatusCode::UNKNOWN_ERROR;
+    }
+
+    // 3. if mount point exist, return MOUNT_POINT_EXIST
+    if (mdsFsInfo.MountPointExist(mountpoint)) {
+        LOG(WARNING) << "MountFs fail, mount point exist, fsName = " << fsName
+                  << ", host = " << mountpoint.host()
+                  << ", mountdir = " << mountpoint.mountdir();
+        return FSStatusCode::MOUNT_POINT_EXIST;
+    }
+
+    // 4. If this is the first mountpoint, init space,
+    if (mdsFsInfo.MountPointEmpty()) {
+        FsInfo tempFsInfo;
+        mdsFsInfo.ConvertToProto(&tempFsInfo);
+        ret = spaceClient_->InitSpace(tempFsInfo);
```

Comment:
lock in fsstorage can't protect concurrent problem.
if two mount operations reached, it can init space twice, and it will lose one mount point record.

Reply:
Consider this problem in next version. Maybe space allocate module should support init space twice, and mount point record  is not important. I think space allocate can ignore mount point information.